### PR TITLE
Reverse MapFrom

### LINKED ISSUE #2067: ReverseMap() - Working with complex types
Like mentioned in #359 I think the default behavior of ReverseMap() is not really obvious. Right now it only works bidirectional for mappings that are implicit like so:

```C#
class A {
 string Member { get; set; }
}

class B {
 string Member { get; set; }
}

cfg.CreateMap<A, B>() .ReverseMap();
```

In this scenario where both classes have members that are named equally ReverseMapping works perfect. But if classes contain members that are different and not mapped via the default (per naming) mapping this doesn´t work properly.

```C#
class A {
 string Member { get; set; }
 string OtherMember { get; set; }
}

class B {
 string Member { get; set; }
 string NonMember { get; set; }
}

cfg.CreateMap<A, B>() .ReverseMap();
```
The way to go to make this work in one direction is to ad the Mapping via `ForMember`.

```C#
cfg.CreateMap<A, B>()
    .ForMember(x => x.NonMember, m => m.MapFrom(a => a.OtherMember))
    .ReverseMap();
```

This now works properly in one direction, but adding `ReverseMap()` doesn´t make it work in both directions so you have to make the mapping similar to this:

```C#
cfg.CreateMap<A, B>()
    .ForMember(x => x.NonMember, m => m.MapFrom(a => a.OtherMember))
    .ReverseMap()
            .ForMember(x => x.OtherMember, m => m.MapFrom(a => a.NonMember));
```
So basically this might lead to some issues. First of all this is code duplication and will duplication will always be error prone. Next I think the default behavior of `ReverseMap()` is not really clear for everyone that uses AutoMapper. 

With this issue I wanted to state a few questions:

* Why is this behavior not implemented? Are there any specific reasons that are not obvious for us right now? Would be great to here about those reasons because I can´t imagine an issue with this right now.

* How can this behavior be achieved? I think there are 3 possible solutions for this:
  1. Add this behavior to `ReverseMap()` per Default (or via config, or param)
  2. Add another Method like `ForMember` for example `MapFromWithReverse`
  3. Add another Method like `MapFrom`  for example `MapFromWithReverse`

Maybe we can use this issue to discuss the if we should, or if we should not implement this behavior and also if we should do so how´s the best way to get this to work.

### PR DESCRIPTION
**Pull request by lbargaoanu**
Fixes #2067.

### CODE REVIEW

File: src/AutoMapper/IMappingExpression.cs [hunk 1]
```diff
@@ -440,4 +441,16 @@ public interface IMappingExpression<TSource, TDestination>
...
+    /// <summary>
+    /// Specify what to reverse.
+    /// </summary>
+    public enum ReverseOptions
+    {
+        Default,
+        /// <summary>
+        /// Reverse simple member accesses MapFrom-s.
+        /// </summary>
+        MapFroms
```
Comment (jbogard) [inline src/AutoMapper/IMappingExpression.cs]: Why do you need this option at all? When would you not want this to happen?

Comment (lbargaoanu) [inline src/AutoMapper/IMappingExpression.cs]: It would break existing code.

Comment (jbogard) [inline src/AutoMapper/IMappingExpression.cs]: Because today those value would just be skipped?

Comment (lbargaoanu) [inline src/AutoMapper/IMappingExpression.cs]: Yes, that's what I meant.

**Comment by jbogard** [comment]
Hmmm so the question is SHOULD we break it if that's the behavior people
actually want.

**Comment by lbargaoanu** [comment]
Well, it's easy for me to say, yes, let's break it, after all the code will be cleaner. And surely no matter what some people will not be happy. I guess the question is how many :)

**Comment by jbogard** [comment]
ah, people are always unhappy!

**Comment by lbargaoanu** [comment]
I guess otherwise they wouldn't get any work done :) So, get rid of the enum and change the default behavior?

**Comment by lbargaoanu** [comment]
I've tried to handle the default flattening. The code is a little silly. It just checks that more than one member is involved and it ignores extension methods. But the fact that the tests pass might mean that it's just good enough :)

**Comment by jbogard** [comment]
And yep, let's get rid of the enum.

**Comment by lbargaoanu** [comment]
Done.

File: src/AutoMapper/PropertyMap.cs [hunk 2]
```diff
@@ -37,7 +39,7 @@ public PropertyMap(PropertyMap inheritedMappedProperty, TypeMap typeMap)
 
         public Type DestinationPropertyType => DestinationProperty.GetMemberType();
 
-        public IEnumerable<MemberInfo> SourceMembers => _memberChain;
+        public IList<MemberInfo> SourceMembers => _memberChain;
```
Comment (jbogard) [inline src/AutoMapper/PropertyMap.cs]: What's the motivation to make it a list instead of enumerable?

Comment (jbogard) [inline src/AutoMapper/PropertyMap.cs]: What's the motivation to make it a list instead of enumerable?

Comment (lbargaoanu) [inline src/AutoMapper/PropertyMap.cs]: I'm checking the count. If it's greater than one, I assume it's the default flattening.

Comment (lbargaoanu) [inline src/AutoMapper/PropertyMap.cs]: A readonly interface might work.

**Pull request merged**